File: src/main/java/net/maple/packets/ItemPackets.java
package net.maple.packets;

import client.inventory.slots.ItemSlot;
import client.inventory.slots.ItemSlotBundle;
import client.inventory.slots.ItemSlotEquip;
import client.inventory.slots.ItemSlotPet;
import constants.ItemConstants;
import util.packet.PacketWriter;

public class ItemPackets {

    public static void encode(ItemSlot item, PacketWriter pw) {
        if (item instanceof ItemSlotEquip) {
            encode((ItemSlotEquip) item, pw);
        } else if (item instanceof ItemSlotBundle) {
            encode((ItemSlotBundle) item, pw);
        } else if (item instanceof ItemSlotPet) {
            encode((ItemSlotPet) item, pw);
        }
    }

    private static void encodeBase(ItemSlot item, PacketWriter pw) {
        pw.writeInt(item.getTemplateId());
        pw.writeBool(item.getCashItemSN() > 0);
        if (item.getCashItemSN() > 0) {
            pw.writeLong(item.getCashItemSN());
        }
        pw.writeLong(item.getExpire() > 0 ? item.getExpire() : 0);
    }

    private static void encode(ItemSlotEquip item, PacketWriter pw) {
        pw.writeByte((byte) 1);
        encodeBase(item, pw);
        pw.writeByte(item.getRUC());
        pw.writeByte(item.getCUC());

        pw.writeShort(item.getSTR());
        pw.writeShort(item.getDEX());
        pw.writeShort(item.getINT());
        pw.writeShort(item.getLUK());
        pw.writeShort(item.getMaxHP());
        pw.writeShort(item.getMaxMP());
        pw.writeShort(item.getPAD());
        pw.writeShort(item.getMAD());
        pw.writeShort(item.getPDD());
        pw.writeShort(item.getMDD());
        pw.writeShort(item.getACC());
        pw.writeShort(item.getEVA());

        pw.writeShort(item.getCraft());
        pw.writeShort(item.getSpeed());
        pw.writeShort(item.getJump());
        pw.writeMapleString(item.getTitle());
        pw.writeShort(item.getAttribute());

        pw.writeByte(item.getLevelUpType());
        pw.writeByte(item.getLevel());
        pw.writeInt(item.getEXP());
        pw.writeInt(item.getDurability());

        pw.writeByte(item.getGrade());
        pw.writeByte(item.getCHUC());

        pw.writeShort(item.getOption1());
        pw.writeShort(item.getOption2());
        pw.writeShort(item.getOption3());
        pw.writeShort(item.getSocket1());
        pw.writeShort(item.getSocket2());

        if (item.getCashItemSN() == 0) pw.writeLong(0);
        pw.writeLong(0);
        pw.writeInt(0);
    }

    private static void encode(ItemSlotBundle item, PacketWriter pw) {
        pw.writeByte((byte) 2);
        encodeBase(item, pw);

        pw.writeShort(item.getNumber());
        pw.writeMapleString(item.getTitle());
        pw.writeShort(item.getAttribute());

        if (ItemConstants.isRechargeableItem(item.getTemplateId())) {
            pw.writeLong(0);
        }
    }

    public static void encode(ItemSlotPet item, PacketWriter pw) {
        pw.writeByte((byte) 3);
        encodeBase(item, pw);

        pw.writeString(item.getPetName());
        pw.fill(0x00, 13 - item.getPetName().length());

        pw.writeByte(item.getLevel());
        pw.writeShort(item.getTameness());
        pw.writeByte(item.getRepleteness());
        pw.writeLong(item.getDateDead());

        pw.writeShort(item.getPetAttribute());
        pw.writeShort(item.getPetSkill());
        pw.writeInt(item.getRemainLife());
        pw.writeShort(item.getAttribute());
    }
}

File: src/main/java/net/maple/packets/ItemPackets.kt
package net.maple.packets

import client.inventory.item.slots.ItemSlot
import client.inventory.item.slots.ItemSlotBundle
import client.inventory.item.slots.ItemSlotEquip
import client.inventory.item.slots.ItemSlotPet
import constants.ItemConstants.isRechargeableItem
import constants.PacketConstants
import util.packet.PacketWriter

object ItemPackets {

    private const val ENCODE_EQUIP = 0x01
    private const val ENCODE_BUNDLE = 0x02
    private const val ENCODE_PET = 0x03

    fun ItemSlot.encode(pw: PacketWriter) {
        when (this) {
            is ItemSlotEquip -> {
                this.encode(pw)
            }
            is ItemSlotBundle -> {
                this.encode(pw)
            }
            is ItemSlotPet -> {
                this.encode(pw)
            }
        }
    }

    private fun ItemSlot.encodeBase(pw: PacketWriter) {
        pw.writeInt(templateId)
        pw.writeBool(cashItemSN > 0)

        if (cashItemSN > 0) {
            pw.writeLong(cashItemSN)
        }

        pw.writeLong(if (expire > 0) expire else PacketConstants.permanent)
    }

    private fun ItemSlotEquip.encode(pw: PacketWriter) {
        pw.write(ENCODE_EQUIP)
        encodeBase(pw)
        pw.writeByte(ruc)
        pw.writeByte(cuc)

        pw.writeShort(str)
        pw.writeShort(dex)
        pw.writeShort(int)
        pw.writeShort(luk)
        pw.writeShort(maxHP)
        pw.writeShort(maxMP)
        pw.writeShort(pad)
        pw.writeShort(mad)
        pw.writeShort(pdd)
        pw.writeShort(mdd)
        pw.writeShort(acc)
        pw.writeShort(eva)

        pw.writeShort(craft)
        pw.writeShort(speed)
        pw.writeShort(jump)
        pw.writeMapleString(title)
        pw.writeShort(attribute)

        pw.writeByte(levelUpType)
        pw.writeByte(level)
        pw.writeInt(exp)
        pw.writeInt(durability)

        pw.writeInt(iuc)

        pw.writeByte(grade)
        pw.writeByte(chuc)

        pw.writeShort(option1)
        pw.writeShort(option2)
        pw.writeShort(option3)
        pw.writeShort(0) // socket1
        pw.writeShort(0) // socket2

        if (cashItemSN == 0L) pw.writeLong(0)
        pw.writeLong(0)
        pw.writeInt(0)
    }

    private fun ItemSlotBundle.encode(pw: PacketWriter) {
        pw.write(ENCODE_BUNDLE)
        encodeBase(pw)

        pw.writeShort(number)
        pw.writeMapleString(title)
        pw.writeShort(attribute)

        if (isRechargeableItem(templateId)) {
            pw.writeLong(0)
        }
    }

    private fun ItemSlotPet.encode(pw: PacketWriter) {
        pw.write(ENCODE_PET)
        encodeBase(pw)

        pw.writeFixedString(petName, 13)
        pw.writeByte(level)
        pw.writeShort(tameness)
        pw.writeByte(repleteness)
        pw.writeLong(dateDead)

        pw.writeShort(petAttribute)
        pw.writeShort(petSkill)
        pw.writeInt(remainLife)
        pw.writeShort(attribute)
    }
}